storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+		if (sdio_cmdSend(host, SDIO_ACMD6_SET_BUS_WIDTH, 2, NULL) < 0) {
+			LOG_ERROR("bus widening failed");
+			return -EIO;
+		}
+
+		*(host->base + SDHOST_REG_HOST_CONTROL) |= HOST_CONTROL_4_BIT_MODE;
+		usleep(10);
+		if (sdio_cmdSendEx(host, SDIO_ACMD13_SD_STATUS, 0, NULL, 0, bigRegs) < 0) {
+			LOG_ERROR("bus widening failed");
+			return -EIO;
+		}
+
+		if (SD_STATUS_DAT_BUS_WIDTH(bigRegs) != SD_STATUS_DAT_BUS_WIDTH_4_BIT) {
+			LOG_ERROR("bus widening failed");
+			return -EIO;
+		}
+	}
+
+	bool isHighSpeedSupported = false;
+	if (cmd6Supported) {
+		switchFunctionArg = SDIO_SWITCH_FUNC_GET | SDIO_SWITCH_FUNC_HIGH_SPEED;
+		if (sdio_cmdSendEx(host, SDIO_CMD6_SWITCH_FUNC, switchFunctionArg, NULL, 0, bigRegs) == 0) {
+			accessModeFunctions = sdcard_extractFunctionGroupInfo(bigRegs, SDCARD_FUNCTION_GROUP_ACCESS_MODE);
+			if ((accessModeFunctions & SDCARD_FUNCTION_GROUP_ACCESS_MODE_HIGH_SPEED) != 0) {
+				switchFunctionArg |= SDIO_SWITCH_FUNC_SET;
+				if (sdio_cmdSendEx(host, SDIO_CMD6_SWITCH_FUNC, switchFunctionArg, NULL, 0, bigRegs) == 0) {
+					if (sdcard_extractFunctionSwitchResult(bigRegs, SDCARD_FUNCTION_GROUP_ACCESS_MODE) == 1) {
+						isHighSpeedSupported = true;
+					}
+				}
+			}
+		}
+	}
+
+	if (isHighSpeedSupported) {
+		TRACE("using HS mode");
+		if (sdcard_configClockAndPower(host, SD_FREQ_50M) < 0) {
+			return -EIO;
+		}
+	}
+	else {
+		TRACE("HS mode not supported");
+		if (sdcard_configClockAndPower(host, SD_FREQ_25M) < 0) {
+			return -EIO;
+		}
+	}
+
+	usleep(10);
+
+	/* Perform a final transaction to check if data transfer is working (we only care about retcode this time) */
+	if (sdio_cmdSendEx(host, SDIO_ACMD13_SD_STATUS, 0, NULL, 0, NULL) < 0) {
+		LOG_ERROR("bus speed change failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+
+static int sdcard_startEventISR(sdcard_hostData_t *host, int interruptNum)
+{
+	if (mutexCreate(&host->eventLock) < 0) {
+		return -ENOMEM;
+	}
+
+	if (condCreate(&host->eventCond) < 0) {
+		resourceDestroy(host->eventLock);
+		return -ENOMEM;
+	}
+
+	interrupt(interruptNum, sdhost_isr, host, host->eventCond, &host->isrHandle);
+
+	return 0;
+}
+
+
+static void _sdcard_free(sdcard_hostData_t *host)
+{
+	TRACE("freeing resources");
+	sdhost_reset(host, CLOCK_CONTROL_RESET_ALL);
+	host->sdioInitialized = false;
+
+	if (host->cmdLock != 0) {
+		resourceDestroy(host->cmdLock);
+		host->cmdLock = -1;
+	}
+
+	if (host->dmaBuffer != NULL) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		host->dmaBufferPhys = (addr_t)NULL;
+	}
+
+	*(host->base + SDHOST_REG_INTR_STATUS_ENABLE) = 0;
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) = 0;
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) = 0;
+}
+
+
+void sdcard_free(unsigned int slot)
+{
+	GET_HOST_OR_RETURN();
+	if (!host->sdioInitialized) {
+		return;
+	}
+
+	mutexLock(host->eventLock);
+	_sdcard_free(host);
+	mutexUnlock(host->eventLock);
+	resourceDestroy(host->eventLock);
```

Comment:
You can destroy host->isrHandle handle too to deregister the irq

Reply:
[Done](https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/2ef22535ad0bfe6a556013d37ac638bf443160ac/storage/zynq7000-sdcard/sdcard.c#L717)